File: examples/imported-agents/langgraph-agenthub-js/demo_graph.js
/**
 * A small LangGraph.js graph, standing in for the one a company already has.
 *
 * Nothing in this file knows about the Agents Hub, and that is the point: it is
 * ordinary LangGraph code of the shape most teams write, and `adapter.js` serves
 * it without a single line of it changing. Read it as "their repository".
 *
 * It has a conditional edge on purpose. A straight line of nodes would not show
 * what the hub's live view is actually for: watching which branch a run took.
 *
 * The model is real when a provider key is present and a deterministic fake
 * otherwise, so the example runs, streams and can be demoed offline.
 *
 * Written as JavaScript with no build step so it can be read and run as it is.
 * The TypeScript version of this file is the same code with types on it: the
 * adapter loads either, since by then it is whatever your own toolchain runs.
 */
import { tool } from "@langchain/core/tools";
import { END, START, StateGraph } from "@langchain/langgraph";
import { MessagesAnnotation } from "@langchain/langgraph";
import { z } from "zod";

export const lookupPricing = tool(
  async ({ product }) => {
    const prices = { starter: "$29/mo", team: "$99/mo", enterprise: "talk to sales" };
    const key = String(product ?? "").trim().toLowerCase();
    return prices[key] ?? `no price on file for '${product}'`;
  },
  {
    name: "lookup_pricing",
    description: "Look up the list price of a product.",
    schema: z.object({ product: z.string() }),
  },
);

let model = null;

/**
 * A real chat model when configured, a streaming fake otherwise.
 *
 * Cached so the fake's scripted replies advance across nodes instead of every
 * node getting the first line again, and so a real model is constructed once.
 */
async function chatModel() {
  if (model) return model;
  if (process.env.OPENAI_API_KEY) {
    const { ChatOpenAI } = await import("@langchain/openai");
    model = new ChatOpenAI({ model: process.env.DEMO_MODEL ?? "gpt-4o-mini", streaming: true });
    return model;
  }
  const { FakeListChatModel } = await import("@langchain/core/utils/testing");
  model = new FakeListChatModel({
    responses: [
      "Reading the question and deciding what is needed.",
      "Here is the answer, based on what the tools returned.",
    ],
  });
  return model;
}

function lastHumanText(messages) {
  for (let i = messages.length - 1; i >= 0; i -= 1) {
    const message = messages[i];
    const role = message?.getType?.() ?? message?.role ?? message?.type;
    if (role === "human" || role === "user") return String(message?.content ?? "").toLowerCase();
  }
  return "";
}

/** First node: think about the request. */
async function triage(state) {
  const reply = await (await chatModel()).invoke(state.messages);
  return { messages: [reply] };
}

/** The branch. Whether this request needs a pricing lookup. */
function needsPricing(state) {
  const text = lastHumanText(state.messages);
  return ["price", "pricing", "cost", "plan"].some((word) => text.includes(word)) ? "pricing" : "answer";
}

/** Tool node: the graph's own tool layer, which the hub never supplies. */
async function pricing(state) {
  let plan = "team";
  for (const candidate of ["starter", "team", "enterprise"]) {
    for (const message of state.messages) {
      if (String(message?.content ?? "").toLowerCase().includes(candidate)) plan = candidate;
    }
  }
  const result = await lookupPricing.invoke({ product: plan });
  return { messages: [{ role: "assistant", content: `Pricing for ${plan}: ${result}` }] };
}

/** Last node: produce the reply the caller sees. */
async function answer(state) {
  const reply = await (await chatModel()).invoke(state.messages);
  return { messages: [reply] };
}

export function build() {
  return new StateGraph(MessagesAnnotation)
    .addNode("triage", triage)
    .addNode("pricing", pricing)
    .addNode("answer", answer)
    .addEdge(START, "triage")
    .addConditionalEdges("triage", needsPricing, { pricing: "pricing", answer: "answer" })
    .addEdge("pricing", "answer")
    .addEdge("answer", END);
}

// What AGENTHUB_GRAPH points at: ./demo_graph.js:graph
export const graph = build().compile();

File: examples/imported-agents/langgraph-agenthub/demo_graph.py
"""A small LangGraph graph, standing in for the one a company already has.

Nothing in this file knows about the Agents Hub, and that is the point: it is
ordinary LangGraph code of the shape most teams write, and ``adapter.py`` serves
it without a single line of it changing. Read it as "their repository".

It has a conditional edge on purpose. A straight line of nodes would not show
what the hub's live view is actually for: watching which branch a run took.

The model is real when a provider key is present and a deterministic fake
otherwise, so the example runs, streams and can be demoed offline.
"""
from __future__ import annotations

import functools
import itertools
import os
from typing import Annotated, TypedDict

from langchain_core.messages import AIMessage
from langchain_core.tools import tool
from langgraph.graph import END, START, StateGraph
from langgraph.graph.message import add_messages


class State(TypedDict):
    messages: Annotated[list, add_messages]


@tool
def lookup_pricing(product: str) -> str:
    """Look up the list price of a product."""
    prices = {"starter": "$29/mo", "team": "$99/mo", "enterprise": "talk to sales"}
    return prices.get(product.strip().lower(), f"no price on file for '{product}'")


@functools.lru_cache(maxsize=1)
def _model():
    """A real chat model when configured, a streaming fake otherwise.

    Cached so the fake's scripted replies advance across nodes instead of every
    node getting the first line again, and so a real model is constructed once.
    """
    if os.environ.get("OPENAI_API_KEY"):
        from langchain_openai import ChatOpenAI
        return ChatOpenAI(model=os.environ.get("DEMO_MODEL", "gpt-4o-mini"), streaming=True)
    if os.environ.get("ANTHROPIC_API_KEY"):
        from langchain_anthropic import ChatAnthropic
        return ChatAnthropic(model=os.environ.get("DEMO_MODEL", "claude-haiku-4-5-20251001"))
    from langchain_core.language_models.fake_chat_models import GenericFakeChatModel
    # cycle() because the graph calls the model more than once per run and a
    # plain iterator would raise StopIteration on the second node.
    return GenericFakeChatModel(messages=itertools.cycle([
        AIMessage(content="Reading the question and deciding what is needed."),
        AIMessage(content="Here is the answer, based on what the tools returned."),
    ]))


def triage(state: State) -> dict:
    """First node: think about the request."""
    reply = _model().invoke(state["messages"])
    return {"messages": [reply]}


def needs_pricing(state: State) -> str:
    """The branch. Whether this request needs a pricing lookup."""
    text = ""
    for message in state["messages"]:
        if getattr(message, "type", "") == "human":
            text = str(getattr(message, "content", "")).lower()
    return "pricing" if any(word in text for word in ("price", "pricing", "cost", "plan")) else "answer"


def pricing(state: State) -> dict:
    """Tool node: the graph's own tool layer, which the hub never supplies."""
    plan = "team"
    for candidate in ("starter", "team", "enterprise"):
        for message in state["messages"]:
            if candidate in str(getattr(message, "content", "")).lower():
                plan = candidate
    result = lookup_pricing.invoke({"product": plan})
    return {"messages": [AIMessage(content=f"Pricing for {plan}: {result}")]}


def answer(state: State) -> dict:
    """Last node: produce the reply the caller sees."""
    reply = _model().invoke(state["messages"])
    return {"messages": [reply]}


def build() -> StateGraph:
    builder = StateGraph(State)
    builder.add_node("triage", triage)
    builder.add_node("pricing", pricing)
    builder.add_node("answer", answer)
    builder.add_edge(START, "triage")
    builder.add_conditional_edges("triage", needs_pricing, {"pricing": "pricing", "answer": "answer"})
    builder.add_edge("pricing", "answer")
    builder.add_edge("answer", END)
    return builder


# What AGENTHUB_GRAPH points at: demo_graph:graph
graph = build().compile()
